# All you need to know in pytorch :-)

An short introduction about PyTorch and about the chosen functions. 

Pytorch is an open source library which contains data structures for deep learning and uses the scripting language as a wrappers for the same.

- arrange

 Returns one dimensional tensor of size  [end-start/step] which has values ranges from start to end with common difference step taken from the start.
- eye

 torch.eye() returns an identity matrix . An identity matrix is nothing but its diagonal value is 1 .

The upper and lower triangle values are zero
- reshape

 torch.reshape() function is used to rearrange elements in the specified size . Inputs are same before and after reshaping . shape only differs after calling function reshape()

- add

  torch.add() adds the scalar to the each element of input .if alpha is specified ,then it will multiply alpha and scalar and add those value to each element in an input

- stack

  Concatenates a sequence of tensors along a new dimension.

  All tensors need to be of the same size

Before we begin, let's install and import PyTorch

In [ ]:
# Uncomment and run the appropriate command for your operating system, if required

# Linux / Binder
# !pip install numpy torch==1.7.0+cpu torchvision==0.8.1+cpu torchaudio==0.7.0 -f https://download.pytorch.org/whl/torch_stable.html

# Windows
# !pip install numpy torch==1.7.0+cpu torchvision==0.8.1+cpu torchaudio==0.7.0 -f https://download.pytorch.org/whl/torch_stable.html

# MacOS
# !pip install numpy torch torchvision torchaudio

In [ ]:
# Import torch and other required modules
import torch

## Function 1 - torch.arange() 

torch.arange(start=0, end, step=1, *, out=None, dtype=None, layout=torch.strided, device=None, requires_grad=False) 



1.   Function torch.arange() returns a 1D tensor of size [end-start/step] with values ranges from start to end with a common difference step
2.  Parameters start -> starting number,
end -> end number ,
step-> gap between each pair of adajacent points ,
out = output tensor , 
requires_grad -> interested in derivation or not


In [ ]:
# Example 1 - working (change this)
torch.arange(10,100,3.90)

tensor([10.0000, 13.9000, 17.8000, 21.7000, 25.6000, 29.5000, 33.4000, 37.3000,
        41.2000, 45.1000, 49.0000, 52.9000, 56.8000, 60.7000, 64.6000, 68.5000,
        72.4000, 76.3000, 80.2000, 84.1000, 88.0000, 91.9000, 95.8000, 99.7000])

Explanation about example

A simple working example to understand how the function arange works . starting number is given as 10 and end as 100 and specified step as 3.90

The vectors displayed contains number starting from number 10 and 3.90 gets added to 10 , then the next number is 13.90.Gap between adajacent number is 3.90 .Displayed all the numbers as a 1D vector starting from 10 with a gap 3.90 and stopped once the number exceeds 100.

size of a vector = [100-10]/3.90 = 24

In [ ]:
torch.arange(10,100,3,dtype=float,requires_grad=True)

tensor([10., 13., 16., 19., 22., 25., 28., 31., 34., 37., 40., 43., 46., 49.,
        52., 55., 58., 61., 64., 67., 70., 73., 76., 79., 82., 85., 88., 91.,
        94., 97.], dtype=torch.float64, requires_grad=True)

Also, specify dtype as float to get the values in torch.float64 datatype

In [ ]:
# Example 3 - breaking (to illustrate when it breaks)
torch.arange(10,100,-1,dtype=float,requires_grad=True)

RuntimeError: ignored

Step value is specified as negative then the lower bound must be greater than upper bound.Careful while using negative step

For eg : torch.arange(10,100,-1,dtype=float,requires_grad=True)

arange () function when it is required to create an array of numbers ranges from [start end) with common gap

Let's save our work using Jovian before continuing.

In [ ]:
!pip install jovian --upgrade --quiet

In [ ]:
import jovian

In [ ]:
jovian.commit(project='01-tensor-operations')

[jovian] Detected Colab notebook...
[jovian] Please enter your API key ( from https://jovian.ai/ ):
API KEY: ··········
[jovian] Uploading colab notebook to Jovian...
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ai/isantha96/01-tensor-operations


'https://jovian.ai/isantha96/01-tensor-operations'

## Function 2 - torch.eye()

torch.eye() returns an identity matrix . 

An identity matrix is nothing but its diagonal value is 1 .

The upper and lower triangle values are zero

In [ ]:
# Example 1 - working
torch.eye(5)

tensor([[1., 0., 0., 0., 0.],
        [0., 1., 0., 0., 0.],
        [0., 0., 1., 0., 0.],
        [0., 0., 0., 1., 0.],
        [0., 0., 0., 0., 1.]])

torch.eye(5) creates an identity matrix of size 5 X 5 .The function takes column as the same value of row if the column isnt specified

In [ ]:
# Example 2 - working
torch.eye(3,4)

tensor([[1., 0., 0., 0.],
        [0., 1., 0., 0.],
        [0., 0., 1., 0.]])

Also,specify columns to have the output of desired size

In [ ]:
# Example 3 - breaking (to illustrate when it breaks)
torch.eye(-1)

RuntimeError: ignored

Matrix dimensions should be positive.Here specified row as negative so it resulted in an error

eye() function is useful when we want to create an identity matrix

In [ ]:
jovian.commit(project='01-tensor-operations')

[jovian] Detected Colab notebook...
[jovian] Uploading colab notebook to Jovian...
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ai/isantha96/01-tensor-operations


'https://jovian.ai/isantha96/01-tensor-operations'

## Function 3 - torch.reshape

torch.reshape() function is used to rearrange elements in the specified size . Inputs are same before and after reshaping . shape only differs after calling function reshape()

In [ ]:
# Example 1 - working
torch.reshape(torch.arange(4),(2,2))


tensor([[0, 1],
        [2, 3]])

torch.arange() function returns 1D vector of size 1X4 . Reshape this 1X4 array as 2 x 2 using reshape function.The inputs are same (0,1,2,3) but the shape alone changes to 2x2 as mentioned in the function .

In [ ]:
# Example 2 - working
torch.reshape(torch.tensor([[1,2,3],[4,5,6],[7,8,9]]),(-1,))

tensor([1, 2, 3, 4, 5, 6, 7, 8, 9])

Also , reshape the 2D vector as 1D vector using reshape function .

First created the 2D tensor vector using torch.tensor function and then reshape it as 1D by specifying (-1,)

In [ ]:
# Example 3 - breaking (to illustrate when it breaks)
torch.reshape(torch.arange(4),(2,3))

RuntimeError: ignored

Specifying dimensions inside reshape functions require little math to know.

The number of elements in matrix is equal to (number of row * number of column)

In our case , the number of elements are 4 , wanted to reshape it as 2x3

2x3 matrix requires 6 elements so it resulted in an error



Obviously , reshape is an useful function . Its used during training ML/DL models . 

In [ ]:
jovian.commit(project='01-tensor-operations')

[jovian] Detected Colab notebook...
[jovian] Uploading colab notebook to Jovian...
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ai/isantha96/01-tensor-operations


'https://jovian.ai/isantha96/01-tensor-operations'

## Function 4 - torch.add()

Adds the scalar other to each element of the input and returns a new resulting tensor.


out=input+alpha * other  (alpha by default is 1)

If input is of type FloatTensor or DoubleTensor, other must be a real number, otherwise it should be an integer.

In [ ]:
# Example 1 - working
torch.add(torch.tensor([[1,2],[3,4]]),4)

tensor([[5, 6],
        [7, 8]])

torch.add() function add the value 4 to the existing matrix 

input : 

[1 2

3  4]

Scalar specified : 4 

alpha : 1

other : Scalar * alpha : 4

output :

[5 6 

7 8 ]



In [ ]:
# Example 2 - working
torch.add(torch.tensor([[1,2],[3,4]]),4,alpha=2)

tensor([[ 9, 10],
        [11, 12]])

torch.add() function add the value (4*2) to the existing matrix 

input : 

[1 2

3  4]

Scalar specified : 4 

alpha : 2

other : scalar*alpha

output :

[9 10 

11 12 ]

In [ ]:
# Example 3 - breaking (to illustrate when it breaks)
torch.add(torch.tensor([[1,2],[3,4]]),torch.tensor([1,2,3]),alpha=2)


RuntimeError: ignored

Adding two matrices require both matrix size to be same dimension

torch.add() function is used when its required add scalar to the matrix

In [ ]:
jovian.commit(project='01-tensor-operations')

[jovian] Detected Colab notebook...
[jovian] Uploading colab notebook to Jovian...
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ai/isantha96/01-tensor-operations


'https://jovian.ai/isantha96/01-tensor-operations'

## Function 5 - torch.stack()

Concatenates a sequence of tensors along a new dimension.

All tensors need to be of the same size

In [ ]:
# Example 1 - working
a= torch.tensor([[1,2],[3,4]])
b = torch.tensor([[5,6],[7,8]])
torch.stack([a,b],dim=2)

tensor([[[1, 5],
         [2, 6]],

        [[3, 7],
         [4, 8]]])

In this example , there are two tensors of size 2x2 . stack() is used to stack two tensors of given dimension

input :

Tensor a :

[1, 2],

[3, 4]

Tensor b :

[5, 6],

[7, 8]

stack a b :

If the dimension is zero , it ll concatenate two matrices.

[[1, 2]

[3, 4]],

[5, 6],

[7, 8]]

If the dimension is one , it ll append rowwise.

first row of a followed by followed by first row of b .second row of a then b

[[[1, 2],

[5, 6]],

[[3, 4],

[7, 8]]]

If the dimension is two , it ll append columnwise



[1 5],

[2 6],

[3,7]

[4,8]






In [ ]:
# Example 2 - working
a= torch.tensor([[1,2],[3,4]])
b = torch.tensor([[5,6],[7,8]])
torch.stack([a,b],dim=0)

tensor([[[1, 2],
         [3, 4]],

        [[5, 6],
         [7, 8]]])

dim = 0 , it concatenates two matrices

In [ ]:
# Example 3 - breaking (to illustrate when it breaks)
a= torch.tensor([[1,2],[3,4]])
b = torch.tensor([[5,6,7],[7,8]])
torch.stack([a,b],dim=2)

ValueError: ignored

Stack requires two matrix dimensions must be same

torch.stack() function is used when its required to create new matrix with two same matrix of new dimension.

In [ ]:
jovian.commit(project='01-tensor-operations')

[jovian] Detected Colab notebook...
[jovian] Uploading colab notebook to Jovian...
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ai/isantha96/01-tensor-operations


'https://jovian.ai/isantha96/01-tensor-operations'

## Conclusion

Summarize what was covered in this notebook, and where to go next

In this notebook, we saw five useful functions in pytorch and explore more functions in the documentation

## Reference Links
Provide links to your references and other interesting articles about tensors
* Official documentation for tensor operations: https://pytorch.org/docs/stable/torch.html
* ...